In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [10]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
numerical_columns = df_market_data.select_dtypes(include=['float64', 'int64']).columns.tolist()

scaler = StandardScaler()

scaled_data = scaler.fit_transform(df_market_data[numerical_columns])

scaled_data


array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [14]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [15]:
# Create a DataFrame with the scaled data
scaled_df = pd.DataFrame(scaled_data, columns=numerical_columns)
# Copy the crypto names from the original DataFrame
scaled_df['coin_id'] = df_market_data.index.values

# Set the coin_id column as index
scaled_df.set_index('coin_id', inplace=True)

# Display the scaled DataFrame
scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [16]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [18]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for k in range(1, 12):
    k_model = KMeans(n_clusters=k, random_state=1) # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model.fit(scaled_df) # 2. Fit the model to the data using `df_market_data_scaled`
    inertia.append(k_model.inertia_) # 3. Append the model.inertia_ to the inertia list

print(inertia)


[287.00000000000006, 212.1233420748626, 165.13675233446813, 79.02243535120977, 66.41305075939438, 61.634088110682214, 46.07446744103022, 43.39269411734828, 32.91322265971526, 30.46911129327036, 26.480101374398934]


In [20]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': list(range(1, 12)),  
    'inertia': inertia  
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

df_elbow

k     inertia
0    1  287.000000
1    2  212.123342
2    3  165.136752
3    4   79.022435
4    5   66.413051
5    6   61.634088
6    7   46.074467
7    8   43.392694
8    9   32.913223
9   10   30.469111
10  11   26.480101

In [ ]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    xticks=list(range(1, 12))  
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Based on the graph and the inertia table, it appears that k = 3 is the best value. The drop from k=2 to k = 3 is not as significant as the drop from k = 3  to k = 4. Similarly, the change from k = 4 to k = 5 is much less significant than that of k = 2 to k = 3, rendering k = 3 the best option.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [24]:
# Initialize the K-Means model using the best value for k
best_k = 3
kmeans_model = KMeans(n_clusters=best_k, random_state=1)

print(kmeans_model)

KMeans(n_clusters=3, random_state=1)


In [25]:
# Fit the K-Means model using the scaled DataFrame
kmeans_model.fit(scaled_df)

KMeans(n_clusters=3, random_state=1)

In [26]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
cluster_labels = kmeans_model.predict(scaled_df)

# Print the resulting array of cluster values.
print("Cluster Labels:")
print(cluster_labels)

Cluster Labels:
[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 0 2 0 0 1
 0 0 0 0]


In [27]:
# Create a copy of the scaled DataFrame
scaled_df_copy = scaled_df.copy()

In [28]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
scaled_df_copy['Predicted_Cluster'] = cluster_labels

# Display the copy of the scaled DataFrame
scaled_df_copy

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [52]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scatter_plot = scaled_df_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Predicted_Cluster",  # Color the points by the predicted clusters
    hover_cols=["crypto_name"],  # Add cryptocurrency name for identification
    title="Cryptocurrency Price Change Percentages",
    xlabel="24h Price Change Percentage",
    ylabel="7d Price Change Percentage"
)
scatter_plot

:NdOverlay   [Predicted_Cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis.

In [31]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [34]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_data = pca.fit_transform(scaled_df)

pca_df = pd.DataFrame(pca_data, columns=["PCA1", "PCA2", "PCA3"])

# View the scaled PCA data
pca_df.head()

PCA1      PCA2      PCA3
0 -0.600667  0.842760  0.461595
1 -0.458261  0.458466  0.952877
2 -0.433070 -0.168126 -0.641752
3 -0.471835 -0.222660 -0.479053
4 -1.157800  2.041209  1.859715

In [35]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
explained_variance

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** In order to get the total variance of the three principal components, we would sum up their individual variances. So, 0.3719856 + 0.34700813 + 0.17603793 = roughly 0.895. This means that these components account for roughly 89.5% of variance.

In [36]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(pca_data, columns=["PCA1", "PCA2", "PCA3"])

# Copy the crypto names from the original scaled DataFrame
pca_df['coin_id'] = scaled_df.index.values

# Set the coin_id column as index
pca_df.set_index('coin_id', inplace=True)

# Display the scaled PCA DataFrame
pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [37]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [38]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    # Create a KMeans model using the loop counter for n_clusters
    model = KMeans(n_clusters=k, random_state=1)
    
    # Fit the model to the data using `df_market_data_pca`
    model.fit(pca_df)
    
    # Append the model's inertia to the inertia list
    inertia.append(model.inertia_)

inertia

[256.8740855678924,
 182.33953007775648,
 135.44240762454208,
 49.66549665179735,
 38.67258217591795,
 34.50757653758153,
 23.72836478127476,
 18.574523440183192,
 15.572672210433337,
 11.407666572096913,
 8.058821888712728]

In [48]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': list(range(1, 12)),  
    'inertia': inertia  
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data)

df_elbow_pca


k     inertia
0    1  256.874086
1    2  182.339530
2    3  135.442408
3    4   49.665497
4    5   38.672582
5    6   34.507577
6    7   23.728365
7    8   18.574523
8    9   15.572672
9   10   11.407667
10  11    8.058822

In [51]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    xticks=k_values
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Based on both the graph and the table, it appears that the best value for k here is k = 4 for the same reasons as why it was k = 3 before. Determined using the elbow method.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, the answer does differ as it was k = 3 previously but k = 4 now.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [41]:
# Initialize the K-Means model using the best value for k
best_k = 4
kmeans_model = KMeans(n_clusters=best_k, random_state=1)

print(kmeans_model)

KMeans(n_clusters=4, random_state=1)


In [42]:
# Fit the K-Means model using the PCA data
kmeans_model.fit(pca_df)

KMeans(n_clusters=4, random_state=1)

In [43]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
crypto_clusters = kmeans_model.predict(pca_df)

# Print the resulting array of cluster values
print(crypto_clusters)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [44]:
# Create a copy of the scaled PCA DataFrame
df_pca_predictions = pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_pca_predictions['StockCluster'] = crypto_clusters  

# Display the copy of the scaled PCA DataFrame
df_pca_predictions.head()

PCA1      PCA2      PCA3  StockCluster
coin_id                                                 
bitcoin      -0.600667  0.842760  0.461595             2
ethereum     -0.458261  0.458466  0.952877             2
tether       -0.433070 -0.168126 -0.641752             0
ripple       -0.471835 -0.222660 -0.479053             0
bitcoin-cash -1.157800  2.041209  1.859715             2

In [ ]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
pca_scatter_plot = df_pca_predictions.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="StockCluster",  
    hover_cols=["Ticker"],  
    title="Cryptocurrency Clusters Based on PCA",
    xlabel="Principal Component 1 (PC1)",
    ylabel="Principal Component 2 (PC2)"
)
pca_scatter_plot

:NdOverlay   [StockCluster]
   :Scatter   [PCA1]   (PCA2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [49]:
# Composite plot to contrast the Elbow curves
composite_plot = df_elbow.hvplot.line(
    x='k',
    y='inertia',
    label='K-Means',
    color='blue'
) * df_elbow_pca.hvplot.line(
    x='k',
    y='inertia',
    label='Agglomerative Clustering',
    color='green'
)

# Add titles and labels
composite_plot.opts(
    title='Comparison of Elbow Curves',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    xticks=list(range(1, 12)),
    legend_position='top_left'
)

composite_plot

:Overlay
   .Curve.K_hyphen_minus_Means     :Curve   [k]   (inertia)
   .Curve.Agglomerative_Clustering :Curve   [k]   (inertia)

In [50]:
# Composite plot to contrast the clusters
composite_plot = pca_scatter_plot * scatter_plot
composite_plot


:Overlay
   .NdOverlay.I  :NdOverlay   [StockCluster]
      :Scatter   [PCA1]   (PCA2)
   .NdOverlay.II :NdOverlay   [Predicted_Cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features tends to give us a better picture seeing as it accounts for most of the variance and seems to do clustering better. It allows also for easier interpretation of the data. That said, it may lead to the loss of certain information so whether or not it is truly better depends on the specific context.